adcm_aio_client/core/requesters.py

```diff
@@ -22,20 +56,136 @@ def as_dict(self: Self) -> dict: ...
+
+        if "results" in data:
+            data = data["results"]
 
-    async def post(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
+        return data
 
-    async def patch(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
 
-    async def delete(self: Self, path: str) -> AsyncGenerator[RequesterResponse]: ...
+STATUS_ERRORS_MAP = {
+    400: BadRequestError,
+    401: UnauthorizedError,
+    404: NotFoundError,
+    409: ConflictError,
+    500: ServerError,
+}
 
 
-class Session: ...
+def convert_exceptions(func: Callable) -> Callable:
+    @wraps(func)
+    async def wrapper(*arg: Iterable, **kwargs: dict) -> httpx.Response:
+        response = await func(*arg, **kwargs)
+        if response.status_code >= 300:
+            raise STATUS_ERRORS_MAP.get(response.status_code, ResponseError)
+
+        return response
+
+    return wrapper
 
 
 class DefaultRequester(Requester):
-    def __init__(self: Self) -> None: ...
+    __slots__ = ("api_root", "client", "retries", "retry_interval", "_credentials")
+
+    def __init__(self: Self, url: str, retries: int = 5, retry_interval: float = 5.0) -> None:
+        self.retries = retries
+        self.retry_interval = retry_interval
+        self.api_root = urljoin(url, "/api/v2/")
+        self.client = httpx.AsyncClient()
+
+    async def login(self: Self, credentials: Credentials) -> Self:
+        login_url = urljoin(self.api_root, "login/")
+
+        try:
+            response = await self._do_request(self.client.post(url=login_url, data=credentials.dict()))
+        except UnauthorizedError as e:
+            raise WrongCredentialsError from e
+
+        if response.status_code != 200:
+            message = f"Authentication error: {response.status_code} for url: {login_url}"
+            raise LoginError(message)
+
+        self._credentials = credentials
+        return self
+
+    async def get(self: Self, path: str, query_params: dict | None = None) -> HTTPXRequesterResponse:
+        return await self.request(method=self.client.get, path=path, params=query_params or {})
+
+    async def post(self: Self, path: str, data: dict) -> HTTPXRequesterResponse:
+        return await self.request(method=self.client.post, path=path, data=data)
+
+    async def patch(self: Self, path: str, data: dict) -> HTTPXRequesterResponse:
+        return await self.request(method=self.client.patch, path=path, data=data)
+
+    async def delete(self: Self, path: str) -> HTTPXRequesterResponse:
+        return await self.request(method=self.client.delete, path=path)
+
+    async def request(self: Self, method: Callable, path: str, **kwargs: dict) -> HTTPXRequesterResponse:
+        url = urljoin(self.api_root, f"{path}/" if not path.endswith("/") else path)
+
+        try:
+            response = await self._do_request(method(url, **kwargs))
+        except UnauthorizedError:
+            await self._reconnect()
```

Comment:
we should do the same for network error, etc

I still suggest making "reconnect" a decorator


and probably name it "retry request" or smt

Reply:
refactored